### PR TITLE
*: cut off duration.fsp in chunk

### LINKED ISSUE #6968: decimals flag in timediff response is incorrect
1. What did you do?

use MySQL-client to connect TiDB with `--column-type-info` argument.

execute:

```
select timediff('2004-12-30 12:00:00.1', cast('2004-12-30 12:00:01' as datetime));
```

2. What did you expect to see?

```
mysql> select timediff('2004-12-30 12:00:00.1', cast('2004-12-30 12:00:01' as datetime));
Field   1:  `timediff('2004-12-30 12:00:00.1', cast('2004-12-30 12:00:01' as datetime))`
Catalog:    `def`
Database:   ``
Table:      ``
Org_table:  ``
Type:       TIME
Collation:  binary (63)
Length:     12
Max_length: 11
Decimals:   1
Flags:      BINARY 


+----------------------------------------------------------------------------+
| timediff('2004-12-30 12:00:00.1', cast('2004-12-30 12:00:01' as datetime)) |
+----------------------------------------------------------------------------+
| -00:00:00.9                                                                |
+----------------------------------------------------------------------------+
1 row in set (0.00 sec)
```
we hope `Decimals:   1`

3. What did you see instead?

```
mysql> select timediff('2004-12-30 12:00:00.1', cast('2004-12-30 12:00:01' as datetime));
Field   1:  `timediff('2004-12-30 12:00:00.1', cast('2004-12-30 12:00:01' as datetime))`
Catalog:    `def`
Database:   ``
Table:      ``
Org_table:  ``
Type:       TIME
Collation:  binary (63)
Length:     60
Max_length: 11
Decimals:   0
Flags:      BINARY 


+----------------------------------------------------------------------------+
| timediff('2004-12-30 12:00:00.1', cast('2004-12-30 12:00:01' as datetime)) |
+----------------------------------------------------------------------------+
| -00:00:00.9                                                                |
+----------------------------------------------------------------------------+
1 row in set (0.00 sec)
```
but got `Decimals:   0`

4. What version of TiDB are you using (`tidb-server -V` or run `select tidb_version();` on TiDB)?

```
Release Version: v2.1.0-beta-1-gc8bae7fa7
Git Commit Hash: c8bae7fa779f70358efa696c06f2319391079c1d
Git Branch: dev-reduce-duration-memory
UTC Build Time: 2018-07-02 03:52:32
GoVersion: go version go1.10.1 linux/amd64
Race Enabled: false
TiKV Min Version: 2.1.0-alpha.1-ff3dd160846b7d1aed9079c389fc188f7f5ea13e
```

### PR DESCRIPTION
**Pull request by lysu**
## What have you changed? (mandatory)

- Remove `fsp` of `types.Duration` from `Chunk` to reduce memory in chunk
- Fix fix time func ret decimal to fix #6968 via 1e12b0e069cca8a8d12a8fcd43b24bd7fec8f167

This PR is "easy way" to #7013 which is more right and more clear in theory, but with big change and be nightmare to both reviewee and reviewer, but I still think it must be done in future.

- Datum's `collation`, `decimal`, `length` is short-cut for some part coding but wrong for design.
- Duration's fsp need be remove with Datum 3 field which need big change Decimal
- Seperate type infer flow with data flow is best.

But this time, we solve chunk in this small PR first. 

## What is the type of the changes? (mandatory)

- Improvement (non-breaking change which is an improvement to an existing feature)
- Bug fix (non-breaking change which fixes an issue)

## How has this PR been tested? (mandatory)

- unit tests
- integration tests
- and I will Port duration mysql_test for test so WIP but ready for review

---
This change is [Reviewable](https://reviewable.io/reviews/pingcap/tidb/7043)

### CODE REVIEW

File: util/chunk/row.go [hunk 1]
```diff
@@ -88,7 +89,8 @@ func (r Row) GetTime(colIdx int) types.Time {
 // GetDuration returns the Duration value with the colIdx.
 func (r Row) GetDuration(colIdx int) types.Duration {
```
Comment (zz-jason) [inline util/chunk/row.go]: Does this change affect the benchmark of `GetDuration`?

File: expression/builtin_time.go [hunk 2]
```diff
@@ -401,7 +401,16 @@ func (c *timeDiffFunctionClass) getFunction(ctx sessionctx.Context, args []Expre
...
+
+	arg0Dec, err := timePrecision(ctx, args[0])
+	if err != nil {
+		return nil, err
+	}
+	arg1Dec, err := timePrecision(ctx, args[1])
+	if err != nil {
+		return nil, err
+	}
+	bf.tp.Decimal = mathutil.Max(arg0Dec, arg1Dec)
```
Comment (zz-jason) [inline expression/builtin_time.go]: this is a bugfix?

File: expression/integration_test.go [hunk 3]
```diff
@@ -1307,13 +1307,13 @@ func (s *testIntegrationSuite) TestTimeBuiltin(c *C) {
 
 	// test time
 	result = tk.MustQuery("select time('2003-12-31 01:02:03')")
-	result.Check(testkit.Rows("01:02:03"))
+	result.Check(testkit.Rows("01:02:03.000000"))
```
Comment (zz-jason) [inline expression/integration_test.go]: this result is different from mysql:
```sql
MySQL(localhost:3306) > select time('2003-12-31 01:02:03');
+-----------------------------+
| time('2003-12-31 01:02:03') |
+-----------------------------+
| 01:02:03                    |
+-----------------------------+
1 row in set (0.00 sec)
```

File: expression/integration_test.go [hunk 4]
```diff
@@ -1307,13 +1307,13 @@ func (s *testIntegrationSuite) TestTimeBuiltin(c *C) {
...
 	result = tk.MustQuery("select time('2003-12-31 01:02:03.000123')")
 	result.Check(testkit.Rows("01:02:03.000123"))
 	result = tk.MustQuery("select time('01:02:03.000123')")
 	result.Check(testkit.Rows("01:02:03.000123"))
 	result = tk.MustQuery("select time('01:02:03')")
-	result.Check(testkit.Rows("01:02:03"))
+	result.Check(testkit.Rows("01:02:03.000000"))
```
Comment (zz-jason) [inline expression/integration_test.go]: ditto

File: util/chunk/row.go (hunk 1 above)
Comment (lysu) [reply util/chunk/row.go]: I will benchmark later, but it seems will allocate more a `types. Duration` than before, but this is just `GetTime` does...

File: expression/builtin_time.go (hunk 2 above)
Comment (lysu) [reply expression/builtin_time.go]: this decimal field is no use before this PR, so it got wrong value but not affect the user, but now if the duration comes from `chunk.row` will need it to compensate missed fsp info.

**Comment by lysu** [comment]
There are still some truncate test cases give the different result to MySQL(even in master branch), https://github.com/pingcap/tidb-test/pull/574/commits/21411bae6f795c51b152383dcd2dd06d4e3c94ea

**Comment by zz-jason** [comment]
LGTM

But we still need some benchmarks to know the exact performance regression.

File: types/etc.go [hunk 5]
```diff
@@ -96,12 +96,18 @@ func IsBinaryStr(ft *FieldType) bool {
...
-	if ft.Collate != charset.CollationBin && (IsTypeChar(ft.Tp) || IsTypeBlob(ft.Tp) || IsTypeVarchar(ft.Tp) || IsTypeUnspecified(ft.Tp)) {
+	if ft.Collate != charset.CollationBin && IsStr(ft) {
 		return true
 	}
 	return false
 }
 
+// IsStr returns a boolean indicating
+// whether the field type is a string type.
+func IsStr(ft *FieldType) bool {
```
Comment (XuHuaiyu) [inline types/etc.go]: s/ IsStr/ IsString

File: util/chunk/row.go [hunk 6]
```diff
@@ -20,6 +20,7 @@ import (
 	"github.com/pingcap/tidb/types"
 	"github.com/pingcap/tidb/types/json"
 	"github.com/pingcap/tidb/util/hack"
+	"time"
```
Comment (XuHuaiyu) [inline util/chunk/row.go]: put this line to line17

File: expression/builtin_time.go [hunk 7]
```diff
@@ -5370,3 +5401,19 @@ func (b *builtinLastDaySig) evalTime(row types.Row) (types.Time, bool, error) {
 	}
 	return ret, false, nil
 }
+
+func timePrecision(ctx sessionctx.Context, expression Expression) (int, error) {
```
Comment (XuHuaiyu) [inline expression/builtin_time.go]: add a comment for this function.

File: expression/builtin_time.go [hunk 8]
```diff
@@ -5370,3 +5401,19 @@ func (b *builtinLastDaySig) evalTime(row types.Row) (types.Time, bool, error) {
...
+	constExp, isConstant := expression.(*Constant)
+	if isConstant && types.IsStr(expression.GetType()) && !isTemporalColumn(expression) {
+		str, isNil, err := constExp.EvalString(ctx, nil)
+		if isNil || err != nil {
+			return 0, errors.Trace(err)
+		}
+		if n := strings.LastIndexByte(str, '.'); n >= 0 {
```
Comment (XuHuaiyu) [inline expression/builtin_time.go]: This check can be replaced by function GetFsp

File: expression/integration_test.go [hunk 9]
```diff
@@ -2663,7 +2663,7 @@ func (s *testIntegrationSuite) TestCompareBuiltin(c *C) {
 	tk.MustExec(`insert into t2 values(1, 1.1, "2017-08-01 12:01:01", "12:01:01", "abcdef", 0b10101)`)
 
 	result = tk.MustQuery("select coalesce(NULL, a), coalesce(NULL, b, a), coalesce(c, NULL, a, b), coalesce(d, NULL), coalesce(d, c), coalesce(NULL, NULL, e, 1), coalesce(f), coalesce(1, a, b, c, d, e, f) from t2")
-	result.Check(testkit.Rows(fmt.Sprintf("1 1.1 2017-08-01 12:01:01 12:01:01 %s 12:01:01 abcdef 21 1", time.Now().In(tk.Se.GetSessionVars().GetTimeZone()).Format("2006-01-02"))))
+	result.Check(testkit.Rows(fmt.Sprintf("1 1.1 2017-08-01 12:01:01 12:01:01.000000 %s 12:01:01 abcdef 21 1", time.Now().In(tk.Se.GetSessionVars().GetTimeZone()).Format("2006-01-02"))))
```
Comment (XuHuaiyu) [inline expression/integration_test.go]: Is this compatible with mysql?

File: types/fsp.go [hunk 10]
```diff
@@ -22,6 +22,8 @@ import (
 )
 
 const (
+	// WaitFillFsp is fsp need fill by caller if they want to use fsp.
+	WaitFillFsp = -2
```
Comment (XuHuaiyu) [inline types/fsp.go]: When will the caller want to us fsp?

File: expression/integration_test.go (hunk 9 above)
Comment (lysu) [reply expression/integration_test.go]: oh~ that's bug...fixed in new commits

File: types/fsp.go (hunk 10 above)
Comment (lysu) [reply types/fsp.go]: - -~ the old way needs GetDuration caller to check whether `result.fsp == WaitFillFsp` than refill Fsp or do nothing...that's buggy.

Now force pass `fillFsp` paramter in `GetDuration` and no longer need this flag. PTAL~thx

File: expression/builtin_time.go (hunk 7 above)
Comment (lysu) [reply expression/builtin_time.go]: - - name is not very suitable, it was renamed and add comments in new commit.

**Comment by lysu** [comment]
a1fb7386ce5c3fe36bc9233f96c93eea1c34ec02 fix https://github.com/pingcap/tidb-test/pull/574/commits/cbeabfbf874b5ed3438c02b70c3066ea5fc64264 and https://github.com/pingcap/tidb-test/pull/574/commits/d53c88580fc5f450a8cf8076aff9593c5cda36bd

**Comment by XuHuaiyu** [comment]
@lysu `DNM` label can be removed?

File: expression/builtin_time.go [hunk 11]
```diff
@@ -5370,3 +5401,16 @@ func (b *builtinLastDaySig) evalTime(row types.Row) (types.Time, bool, error) {
 	}
 	return ret, false, nil
 }
+
+// expressionFsp calculates the fsp from given expression.
+func expressionFsp(ctx sessionctx.Context, expression Expression) (int, error) {
```
Comment (XuHuaiyu) [inline expression/builtin_time.go]: s/ expressionFsp/ getExpressionFsp

File: expression/typeinfer_test.go [hunk 12]
```diff
@@ -1249,7 +1249,7 @@ func (s *testInferTypeSuite) createTestCase4TimeFuncs() []typeInferTestCase {
 		{"addtime(c_timestamp, c_time_d)", mysql.TypeDatetime, charset.CharsetBin, mysql.BinaryFlag, 26, 4},
 		{"addtime(c_timestamp_d, c_time_d)", mysql.TypeDatetime, charset.CharsetBin, mysql.BinaryFlag, 26, 0},
 		{"addtime(c_time, c_time)", mysql.TypeDuration, charset.CharsetBin, mysql.BinaryFlag, 15, 3},
-		{"addtime(c_time_d, c_time)", mysql.TypeDuration, charset.CharsetBin, mysql.BinaryFlag, 15, 0},
+		{"addtime(c_time_d, c_time)", mysql.TypeDuration, charset.CharsetBin, mysql.BinaryFlag, 15, 3},
```
Comment (XuHuaiyu) [inline expression/typeinfer_test.go]: Do the changes in this file be more reasonable or more compatible?

File: types/row.go [hunk 13]
```diff
@@ -42,7 +42,9 @@ type Row interface {
 	GetTime(colIdx int) Time
 
 	// GetDuration returns the Duration value with the colIdx.
-	GetDuration(colIdx int) Duration
+	// fillFsp is needed for refill fsp info if duration came from chunk.Row which is no longer store fsp info.
+	// if caller make sure that data from Datum or only use Duration.Duration properties can pass 0 as fillFsp
```
Comment (XuHuaiyu) [inline types/row.go]: s/ is/ If
add `.` at the end of this comment

Comment (XuHuaiyu) [inline types/row.go]: s/ is/ If
add `.` at the end of this comment

File: types/time.go [hunk 14]
```diff
@@ -812,7 +813,7 @@ func (d Duration) String() string {
 	}
 
 	fmt.Fprintf(&buf, "%02d:%02d:%02d", hours, minutes, seconds)
-	if d.Fsp > 0 {
+	if d.Fsp > 0 && fraction >= 0 {
```
Comment (XuHuaiyu) [inline types/time.go]: add a comment for this check

File: types/datum.go [hunk 15]
```diff
@@ -1019,8 +1019,18 @@ func (d *Datum) convertToMysqlDuration(sc *stmtctx.StatementContext, target *Fie
...
 		if err != nil {
 			return ret, errors.Trace(err)
-		} else if timeStr[0] == '-' {
-			return ret, ErrInvalidTimeFormat.Gen("Incorrect time value '%s'", timeStr)
+		}
+		timeNum, err := d.ToInt64(sc)
+		if err != nil {
+			return ret, errors.Trace(err)
+		}
+		if timeNum > MaxDuration && timeNum < 10000000000 {
```
Comment (XuHuaiyu) [inline types/datum.go]: what if timeNum > 10000000000

File: expression/builtin_time.go [hunk 16]
```diff
@@ -401,7 +401,16 @@ func (c *timeDiffFunctionClass) getFunction(ctx sessionctx.Context, args []Expre
 	arg0FieldTp, arg1FieldTp := args[0].GetType(), args[1].GetType()
 	arg0Tp, arg1Tp := c.getArgEvalTp(arg0FieldTp), c.getArgEvalTp(arg1FieldTp)
 	bf := newBaseBuiltinFuncWithTp(ctx, args, types.ETDuration, arg0Tp, arg1Tp)
-	bf.tp.Decimal = mathutil.Max(arg0FieldTp.Decimal, arg1FieldTp.Decimal)
+
+	arg0Dec, err := expressionFsp(ctx, args[0])
+	if err != nil {
+		return nil, err
```
Comment (XuHuaiyu) [inline expression/builtin_time.go]: errors.Trace(err)

and line 411

File: expression/typeinfer_test.go (hunk 12 above)
Comment (lysu) [reply expression/typeinfer_test.go]: Yes, in previous we didn't infer decimal and always be zero, but in MySQL takes that.

```
mysql> select version();
+---------------------------+
| version()                 |
+---------------------------+
| 5.7.22-0ubuntu18.04.1-log |
+---------------------------+
1 row in set (0.00 sec)

mysql> select addtime(c_time_d, c_time) from t;
Field   1:  `addtime(c_time_d, c_time)`
Catalog:    `def`
Database:   ``
Table:      ``
Org_table:  ``
Type:       TIME
Collation:  binary (63)
Length:     14
Max_length: 0
Decimals:   3
Flags:      BINARY 
0 rows in set (0.00 sec)
```

File: types/time.go (hunk 14 above)
Comment (lysu) [reply types/time.go]: check fraction seems useless....I removed it

File: types/datum.go (hunk 15 above)
Comment (lysu) [reply types/datum.go]: timeNum > MaxDuration and < 10000000000 must be a wrong value..
but timeNum > 10000000000 will try treat as `'YYYYMMDDHHmmss'`

specially timeNum < 0 and timeNum < -10000000000 should NOT treat as `'YYYYMMDDHHmmss'`

https://github.com/mysql/mysql-server/blob/e7586bb9c11953a81c816ca10708f612403edc24/sql-common/my_time.c#L846

**Comment by coocood** [comment]
LGTM

**Comment by lysu** [comment]
@zz-jason 

```
func BenchmarkGetDurationFromChunk(b *testing.B) {
	chk := chunk.NewChunkWithCapacity([]*types.FieldType{{Tp: mysql.TypeDuration}}, 1)
	chk.AppendDuration(0, types.Duration{Duration: 11233*time.Second, Fsp:1})
	row := chk.GetRow(0)
	b.ResetTimer()
	for i := 0; i < b.N; i++ {
		x: = row.GetDuration(0, 1)
                x = x
	}
}
```
for master:

```
BenchmarkGetDurationFromChunk-8   	2000000000	         1.06 ns/op	       0 B/op	       0 allocs/op
PASS

```

for this pr:

```
BenchmarkGetDurationFromChunk-8   	2000000000	         1.03 ns/op	       0 B/op	       0 allocs/op
PASS
```

File: types/datum.go [hunk 17]
```diff
@@ -1019,8 +1019,19 @@ func (d *Datum) convertToMysqlDuration(sc *stmtctx.StatementContext, target *Fie
...
 		if err != nil {
 			return ret, errors.Trace(err)
-		} else if timeStr[0] == '-' {
-			return ret, ErrInvalidTimeFormat.Gen("Incorrect time value '%s'", timeStr)
+		}
+		timeNum, err := d.ToInt64(sc)
+		if err != nil {
+			return ret, errors.Trace(err)
+		}
+		// For For huge numbers(>'0001-00-00 00-00-00') try full DATETIME in ParseDuration.
```
Comment (XuHuaiyu) [inline types/datum.go]: remove one `For`

Comment (lysu) [reply types/datum.go]: Done

**Pull request merged**